### PR TITLE
Improve hook handling in backward

### PR DESCRIPTION
**Pull request by okuta**
I use #6263 

before
22.832 ms

after
21.415 ms

### CODE REVIEW

**Comment by okuta** [comment]
jenkins, test this please.

File: chainer/function_node.py [hunk 1]
```diff
@@ -1122,25 +1123,27 @@ def _backprop(outputs, inputs, grad_required, retain_grad, grads, loss_scale):
...
+            local_hooks = collections.OrderedDict(chainer.get_function_hooks())
+            local_hooks.update(func.local_function_hooks)
+            hooks = local_hooks.values()  # avoid six for performance
+        else:
+            hooks = base_hooks
 
-        in_data = tuple([x.data for x in func.inputs])
-        out_grad_data = tuple(
-            [None if g is None else g.data for g in gys])
+        in_data = [x.data for x in func.inputs]
```
Comment (mitmul) [inline chainer/function_node.py]: Is it OK not to use `array` attribute instead of `data`?

File: chainer/function_node.py [hunk 2]
```diff
@@ -1122,25 +1123,27 @@ def _backprop(outputs, inputs, grad_required, retain_grad, grads, loss_scale):
...
+        out_grad_data = [None if g is None else g.data for g in gys]
```
Comment (mitmul) [inline chainer/function_node.py]: Ditto.

File: chainer/variable.py [hunk 3]
```diff
@@ -1564,18 +1565,19 @@ def add_cand(cand):
         if not target_input_indexes:
             continue
 
-        in_data = tuple([x.data for x in inputs])
-        out_grad_array = tuple(
-            [None if g is None else g.array for g in out_grad])
-        hooks = chainer.get_function_hooks()
+        in_data = [x.data for x in inputs]
```
Comment (mitmul) [inline chainer/variable.py]: Ditto.

File: chainer/function_node.py (hunk 1 above)
Comment (toslunar) [reply chainer/function_node.py]: `x` is an instance of `VariableNode`, which only has `data`.

**Comment by okuta** [comment]
jenkins, test this please.

File: chainer/function_node.py (hunk 1 above)
Comment (mitmul) [reply chainer/function_node.py]: Oops, got it.

**Comment by mitmul** [comment]
The error failed in the Jenkins jobs are irrelevant to this PR.

**Pull request merged**